File: CODIGO A MARTE/01-temperatura-habitat/cpp/control_temperatura.cpp
// Sistema Crítico de Control de Temperatura para Hábitat Marciano
// Implementación en C++ con gestión de recursos RAII y manejo de excepciones

#include <iostream>
#include <vector>
#include <memory>
#include <thread>
#include <chrono>
#include <mutex>
#include <algorithm>
#include <numeric>
#include <iomanip>
#include <stdexcept>
#include <cmath>

// Constantes críticas del sistema
constexpr float TEMP_MIN_HABITAT = 18.0f;  // °C
constexpr float TEMP_MAX_HABITAT = 24.0f;  // °C
constexpr float TEMP_MIN_HUERTA = 20.0f;   // °C
constexpr float TEMP_MAX_HUERTA = 28.0f;   // °C
constexpr float TEMP_CRITICA_MIN = 10.0f;  // °C - Alerta crítica
constexpr float TEMP_CRITICA_MAX = 35.0f;  // °C - Alerta crítica

enum class ZonaHabitat {
    HABITAT,
    HUERTA
};

enum class EstadoSistema {
    NORMAL,
    ADVERTENCIA,
    CRITICO,
    EMERGENCIA
};

class SensorTemperatura {
private:
    uint32_t id_;
    ZonaHabitat zona_;
    float temperatura_;
    bool activo_;
    uint32_t lecturas_fallidas_;
    mutable std::mutex mutex_;

public:
    SensorTemperatura(uint32_t id, ZonaHabitat zona)
        : id_(id), zona_(zona), temperatura_(20.0f), 
          activo_(true), lecturas_fallidas_(0) {}

    float leer_temperatura() {
        std::lock_guard<std::mutex> lock(mutex_);
        
        if (!activo_ || lecturas_fallidas_ >= 3) {
            throw std::runtime_error("Sensor " + std::to_string(id_) + " inactivo o fallido");
        }

        // Simulación de lectura con variación
        float variacion = std::fmod(id_ * 0.1f, 2.0f) - 1.0f;
        temperatura_ += variacion;
        
        return temperatura_;
    }

    void calibrar() {
        std::lock_guard<std::mutex> lock(mutex_);
        lecturas_fallidas_ = 0;
        activo_ = true;
        std::cout << "✓ Sensor " << id_ << " calibrado\n";
    }

    uint32_t get_id() const { return id_; }
    bool esta_activo() const { 
        std::lock_guard<std::mutex> lock(mutex_);
        return activo_; 
    }
    float get_temperatura() const { 
        std::lock_guard<std::mutex> lock(mutex_);
        return temperatura_; 
    }
};

class ActuadorTermico {
private:
    uint32_t id_;
    ZonaHabitat zona_;
    float potencia_;  // 0.0 a 100.0 %
    bool activo_;
    mutable std::mutex mutex_;

public:
    ActuadorTermico(uint32_t id, ZonaHabitat zona)
        : id_(id), zona_(zona), potencia_(0.0f), activo_(true) {}

    void ajustar_potencia(float nueva_potencia) {
        std::lock_guard<std::mutex> lock(mutex_);
        potencia_ = std::clamp(nueva_potencia, 0.0f, 100.0f);
        std::cout << "🔧 Actuador " << id_ << " ajustado a " 
                  << std::fixed << std::setprecision(1) << potencia_ << "% potencia\n";
    }

    void activar() {
        std::lock_guard<std::mutex> lock(mutex_);
        activo_ = true;
        std::cout << "✓ Actuador " << id_ << " activado\n";
    }

    void desactivar() {
        std::lock_guard<std::mutex> lock(mutex_);
        activo_ = false;
        potencia_ = 0.0f;
        std::cout << "⚠ Actuador " << id_ << " desactivado\n";
    }

    uint32_t get_id() const { return id_; }
    bool esta_activo() const { 
        std::lock_guard<std::mutex> lock(mutex_);
        return activo_; 
    }
    float get_potencia() const { 
        std::lock_guard<std::mutex> lock(mutex_);
        return potencia_; 
    }
};

class SistemaControlTemperatura {
private:
    std::vector<std::unique_ptr<SensorTemperatura>> sensores_habitat_;
    std::vector<std::unique_ptr<SensorTemperatura>> sensores_huerta_;
    std::vector<std::unique_ptr<ActuadorTermico>> actuadores_habitat_;
    std::vector<std::unique_ptr<ActuadorTermico>> actuadores_huerta_;
    EstadoSistema estado_;
    mutable std::mutex mutex_;

    float obtener_temperatura_promedio(ZonaHabitat zona) {
        auto& sensores = (zona == ZonaHabitat::HABITAT) ? 
                         sensores_habitat_ : sensores_huerta_;
        
        std::vector<float> lecturas_validas;
        
        for (auto& sensor : sensores) {
            try {
                float temp = sensor->leer_temperatura();
                lecturas_validas.push_back(temp);
            } catch (const std::exception& e) {
                // Sensor fallido, continuar con otros
            }
        }

        if (lecturas_validas.empty()) {
            throw std::runtime_error("No hay sensores operativos");
        }

        float suma = std::accumulate(lecturas_validas.begin(), 
                                     lecturas_validas.end(), 0.0f);
        return suma / lecturas_validas.size();
    }

    EstadoSistema evaluar_estado(float temp_habitat, float temp_huerta) {
        if (temp_habitat < TEMP_CRITICA_MIN || temp_habitat > TEMP_CRITICA_MAX ||
            temp_huerta < TEMP_CRITICA_MIN || temp_huerta > TEMP_CRITICA_MAX) {
            return EstadoSistema::EMERGENCIA;
        } else if (temp_habitat < TEMP_MIN_HABITAT - 2.0f || 
                   temp_habitat > TEMP_MAX_HABITAT + 2.0f ||
                   temp_huerta < TEMP_MIN_HUERTA - 2.0f || 
                   temp_huerta > TEMP_MAX_HUERTA + 2.0f) {
            return EstadoSistema::CRITICO;
        } else if (temp_habitat < TEMP_MIN_HABITAT || temp_habitat > TEMP_MAX_HABITAT ||
                   temp_huerta < TEMP_MIN_HUERTA || temp_huerta > TEMP_MAX_HUERTA) {
            return EstadoSistema::ADVERTENCIA;
        }
        return EstadoSistema::NORMAL;
    }

    float calcular_potencia_necesaria(float temp_actual, float temp_objetivo) {
        float diferencia = temp_objetivo - temp_actual;
        float potencia = std::clamp(std::abs(diferencia) * 10.0f, 0.0f, 100.0f);
        return (diferencia > 0.0f) ? potencia : -potencia;
    }

    void ajustar_actuadores(ZonaHabitat zona, float potencia) {
        auto& actuadores = (zona == ZonaHabitat::HABITAT) ? 
                          actuadores_habitat_ : actuadores_huerta_;
        
        size_t actuadores_activos = std::count_if(actuadores.begin(), actuadores.end(),
            [](const auto& a) { return a->esta_activo(); });
        
        if (actuadores_activos > 0) {
            float potencia_por_actuador = std::abs(potencia) / actuadores_activos;
            
            for (auto& actuador : actuadores) {
                if (actuador->esta_activo()) {
                    actuador->ajustar_potencia(potencia_por_actuador);
                }
            }
        }
    }

    void protocolo_emergencia() {
        std::cout << "\n🚨 PROTOCOLO DE EMERGENCIA ACTIVADO\n";
        std::cout << "   1. Activando todos los actuadores disponibles\n";
        std::cout << "   2. Notificando a control de misión\n";
        std::cout << "   3. Preparando sistemas de respaldo\n";
        
        for (auto& actuador : actuadores_habitat_) {
            actuador->activar();
            actuador->ajustar_potencia(100.0f);
        }
        
        for (auto& actuador : actuadores_huerta_) {
            actuador->activar();
            actuador->ajustar_potencia(100.0f);
        }
    }

    std::string estado_a_string(EstadoSistema estado) const {
        switch (estado) {
            case EstadoSistema::NORMAL: return "NORMAL";
            case EstadoSistema::ADVERTENCIA: return "ADVERTENCIA";
            case EstadoSistema::CRITICO: return "CRITICO";
            case EstadoSistema::EMERGENCIA: return "EMERGENCIA";
            default: return "DESCONOCIDO";
        }
    }

    std::string icono_estado(EstadoSistema estado) const {
        switch (estado) {
            case EstadoSistema::NORMAL: return "✓";
            case EstadoSistema::ADVERTENCIA: return "⚠";
            case EstadoSistema::CRITICO: return "🔴";
            case EstadoSistema::EMERGENCIA: return "🚨";
            default: return "?";
        }
    }

public:
    SistemaControlTemperatura() : estado_(EstadoSistema::NORMAL) {
        // Redundancia: 3 sensores por zona
        for (uint32_t i = 1; i <= 3; ++i) {
            sensores_habitat_.push_back(
                std::make_unique<SensorTemperatura>(i, ZonaHabitat::HABITAT));
        }
        
        for (uint32_t i = 4; i <= 6; ++i) {
            sensores_huerta_.push_back(
                std::make_unique<SensorTemperatura>(i, ZonaHabitat::HUERTA));
        }

        // Redundancia: 2 actuadores por zona
        for (uint32_t i = 1; i <= 2; ++i) {
            actuadores_habitat_.push_back(
                std::make_unique<ActuadorTermico>(i, ZonaHabitat::HABITAT));
        }
        
        for (uint32_t i = 3; i <= 4; ++i) {
            actuadores_huerta_.push_back(
                std::make_unique<ActuadorTermico>(i, ZonaHabitat::HUERTA));
        }
    }

    void ciclo_control() {
        std::lock_guard<std::mutex> lock(mutex_);
        
        std::cout << "\n" << std::string(60, '=') << "\n";
        std::cout << "🔄 Iniciando ciclo de control de temperatura\n";
        std::cout << std::string(60, '=') << "\n";

        try {
            // Leer temperaturas
            float temp_habitat = obtener_temperatura_promedio(ZonaHabitat::HABITAT);
            float temp_huerta = obtener_temperatura_promedio(ZonaHabitat::HUERTA);

            std::cout << "\n📊 LECTURAS ACTUALES:\n";
            std::cout << std::fixed << std::setprecision(2);
            std::cout << "   Hábitat: " << temp_habitat << "°C (Rango: " 
                      << TEMP_MIN_HABITAT << "°C - " << TEMP_MAX_HABITAT << "°C)\n";
            std::cout << "   Huerta:  " << temp_huerta << "°C (Rango: " 
                      << TEMP_MIN_HUERTA << "°C - " << TEMP_MAX_HUERTA << "°C)\n";

            // Evaluar estado
            estado_ = evaluar_estado(temp_habitat, temp_huerta);
            std::cout << "\n" << icono_estado(estado_) << " ESTADO DEL SISTEMA: " 
                      << estado_a_string(estado_) << "\n";

            // Calcular y aplicar ajustes
            float temp_objetivo_habitat = (TEMP_MIN_HABITAT + TEMP_MAX_HABITAT) / 2.0f;
            float temp_objetivo_huerta = (TEMP_MIN_HUERTA + TEMP_MAX_HUERTA) / 2.0f;

            float potencia_habitat = calcular_potencia_necesaria(temp_habitat, temp_objetivo_habitat);
            float potencia_huerta = calcular_potencia_necesaria(temp_huerta, temp_objetivo_huerta);

            std::cout << "\n🎯 AJUSTES CALCULADOS:\n";
            ajustar_actuadores(ZonaHabitat::HABITAT, potencia_habitat);
            ajustar_actuadores(ZonaHabitat::HUERTA, potencia_huerta);

            // Protocolo de emergencia
            if (estado_ == EstadoSistema::EMERGENCIA) {
                protocolo_emergencia();
            }

            std::cout << "✓ Ciclo completado exitosamente\n";

        } catch (const std::exception& e) {
            std::cerr << "✗ Error en ciclo: " << e.what() << "\n";
            throw;
        }
    }

    void diagnostico_sistema() const {
        std::lock_guard<std::mutex> lock(mutex_);
        
        std::cout << "\n" << std::string(60, '=') << "\n";
        std::cout << "🔍 DIAGNÓSTICO DEL SISTEMA\n";
        std::cout << std::string(60, '=') << "\n";
        
        std::cout << "\n📡 Sensores Hábitat:\n";
        for (const auto& sensor : sensores_habitat_) {
            std::string estado = sensor->esta_activo() ? "✓ Operativo" : "✗ Inactivo";
            std::cout << "   Sensor " << sensor->get_id() << ": " << estado 
                      << " - " << std::fixed << std::setprecision(2) 
                      << sensor->get_temperatura() << "°C\n";
        }
        
        std::cout << "\n📡 Sensores Huerta:\n";
        for (const auto& sensor : sensores_huerta_) {
            std::string estado = sensor->esta_activo() ? "✓ Operativo" : "✗ Inactivo";
            std::cout << "   Sensor " << sensor->get_id() << ": " << estado 
                      << " - " << std::fixed << std::setprecision(2) 
                      << sensor->get_temperatura() << "°C\n";
        }
        
        std::cout << "\n🔧 Actuadores Hábitat:\n";
        for (const auto& actuador : actuadores_habitat_) {
            std::string estado = actuador->esta_activo() ? "✓ Operativo" : "✗ Inactivo";
            std::cout << "   Actuador " << actuador->get_id() << ": " << estado 
                      << " - " << std::fixed << std::setprecision(1) 
                      << actuador->get_potencia() << "% potencia\n";
        }
        
        std::cout << "\n🔧 Actuadores Huerta:\n";
        for (const auto& actuador : actuadores_huerta_) {
            std::string estado = actuador->esta_activo() ? "✓ Operativo" : "✗ Inactivo";
            std::cout << "   Actuador " << actuador->get_id() << ": " << estado 
                      << " - " << std::fixed << std::setprecision(1) 
                      << actuador->get_potencia() << "% potencia\n";
        }
    }
};

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "🚀 SISTEMA DE CONTROL DE TEMPERATURA - HÁBITAT MARCIANO\n";
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Versión: 1.0.0 - C++\n";
    std::cout << "Estado: Sistema Crítico - Máxima Prioridad\n";

    try {
        auto sistema = std::make_unique<SistemaControlTemperatura>();
        
        // Diagnóstico inicial
        sistema->diagnostico_sistema();
        
        std::cout << "\n🔄 Iniciando monitoreo continuo...\n";
        
        // Simulación de 5 ciclos de control
        for (int ciclo = 1; ciclo <= 5; ++ciclo) {
            std::cout << "\n" << std::string(60, '=') << "\n";
            std::cout << "CICLO #" << ciclo << "\n";
            
            std::this_thread::sleep_for(std::chrono::seconds(2));
            
            sistema->ciclo_control();
        }
        
        // Diagnóstico final
        sistema->diagnostico_sistema();
        
        std::cout << "\n" << std::string(60, '=') << "\n";
        std::cout << "✓ Sistema de control finalizado\n";
        std::cout << std::string(60, '=') << "\n";

    } catch (const std::exception& e) {
        std::cerr << "🚨 Error crítico: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

// Made with Bob

File: CODIGO A MARTE/01-temperatura-habitat/python/control_temperatura.py
#!/usr/bin/env python3
"""
Sistema Crítico de Control de Temperatura para Hábitat Marciano
Implementación en Python con type hints y manejo robusto de errores
"""

import time
import threading
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional
from statistics import mean
import sys

# Constantes críticas del sistema
TEMP_MIN_HABITAT = 18.0  # °C
TEMP_MAX_HABITAT = 24.0  # °C
TEMP_MIN_HUERTA = 20.0   # °C
TEMP_MAX_HUERTA = 28.0   # °C
TEMP_CRITICA_MIN = 10.0  # °C - Alerta crítica
TEMP_CRITICA_MAX = 35.0  # °C - Alerta crítica


class ZonaHabitat(Enum):
    """Zonas del hábitat marciano"""
    HABITAT = "habitat"
    HUERTA = "huerta"


class EstadoSistema(Enum):
    """Estados posibles del sistema de control"""
    NORMAL = "normal"
    ADVERTENCIA = "advertencia"
    CRITICO = "critico"
    EMERGENCIA = "emergencia"


@dataclass
class SensorTemperatura:
    """Sensor de temperatura con redundancia y detección de fallos"""
    id: int
    zona: ZonaHabitat
    temperatura: float = 20.0
    activo: bool = True
    lecturas_fallidas: int = 0
    _lock: threading.Lock = field(default_factory=threading.Lock, repr=False)

    def leer_temperatura(self) -> float:
        """Lee la temperatura del sensor con manejo de errores"""
        with self._lock:
            if not self.activo or self.lecturas_fallidas >= 3:
                raise RuntimeError(f"Sensor {self.id} inactivo o fallido")
            
            # Simulación de lectura con variación
            variacion = (self.id * 0.1) % 2.0 - 1.0
            self.temperatura += variacion
            
            return self.temperatura

    def calibrar(self) -> None:
        """Calibra el sensor y resetea contadores de fallo"""
        with self._lock:
            self.lecturas_fallidas = 0
            self.activo = True
            print(f"✓ Sensor {self.id} calibrado")

    def __str__(self) -> str:
        estado = "✓ Operativo" if self.activo else "✗ Inactivo"
        return f"Sensor {self.id}: {estado} - {self.temperatura:.2f}°C"


@dataclass
class ActuadorTermico:
    """Actuador térmico para control de temperatura"""
    id: int
    zona: ZonaHabitat
    potencia: float = 0.0  # 0.0 a 100.0 %
    activo: bool = True
    _lock: threading.Lock = field(default_factory=threading.Lock, repr=False)

    def ajustar_potencia(self, nueva_potencia: float) -> None:
        """Ajusta la potencia del actuador"""
        with self._lock:
            self.potencia = max(0.0, min(100.0, nueva_potencia))
            print(f"🔧 Actuador {self.id} ajustado a {self.potencia:.1f}% potencia")

    def activar(self) -> None:
        """Activa el actuador"""
        with self._lock:
            self.activo = True
            print(f"✓ Actuador {self.id} activado")

    def desactivar(self) -> None:
        """Desactiva el actuador"""
        with self._lock:
            self.activo = False
            self.potencia = 0.0
            print(f"⚠ Actuador {self.id} desactivado")

    def __str__(self) -> str:
        estado = "✓ Operativo" if self.activo else "✗ Inactivo"
        return f"Actuador {self.id}: {estado} - {self.potencia:.1f}% potencia"


class SistemaControlTemperatura:
    """Sistema principal de control de temperatura del hábitat marciano"""

    def __init__(self):
        # Redundancia: 3 sensores por zona
        self.sensores_habitat: List[SensorTemperatura] = [
            SensorTemperatura(id=i, zona=ZonaHabitat.HABITAT)
            for i in range(1, 4)
        ]
        
        self.sensores_huerta: List[SensorTemperatura] = [
            SensorTemperatura(id=i, zona=ZonaHabitat.HUERTA)
            for i in range(4, 7)
        ]

        # Redundancia: 2 actuadores por zona
        self.actuadores_habitat: List[ActuadorTermico] = [
            ActuadorTermico(id=i, zona=ZonaHabitat.HABITAT)
            for i in range(1, 3)
        ]
        
        self.actuadores_huerta: List[ActuadorTermico] = [
            ActuadorTermico(id=i, zona=ZonaHabitat.HUERTA)
            for i in range(3, 5)
        ]

        self.estado: EstadoSistema = EstadoSistema.NORMAL
        self._lock = threading.Lock()

    def obtener_temperatura_promedio(self, zona: ZonaHabitat) -> float:
        """Obtiene la temperatura promedio de una zona usando sensores redundantes"""
        sensores = (self.sensores_habitat if zona == ZonaHabitat.HABITAT 
                   else self.sensores_huerta)
        
        lecturas_validas = []
        
        for sensor in sensores:
            try:
                temp = sensor.leer_temperatura()
                lecturas_validas.append(temp)
            except RuntimeError:
                # Sensor fallido, continuar con otros
                pass
        
        if not lecturas_validas:
            raise RuntimeError("No hay sensores operativos")
        
        return mean(lecturas_validas)

    def evaluar_estado(self, temp_habitat: float, temp_huerta: float) -> EstadoSistema:
        """Evalúa el estado del sistema basado en las temperaturas"""
        if (temp_habitat < TEMP_CRITICA_MIN or temp_habitat > TEMP_CRITICA_MAX or
            temp_huerta < TEMP_CRITICA_MIN or temp_huerta > TEMP_CRITICA_MAX):
            return EstadoSistema.EMERGENCIA
        elif (temp_habitat < TEMP_MIN_HABITAT - 2.0 or temp_habitat > TEMP_MAX_HABITAT + 2.0 or
              temp_huerta < TEMP_MIN_HUERTA - 2.0 or temp_huerta > TEMP_MAX_HUERTA + 2.0):
            return EstadoSistema.CRITICO
        elif (temp_habitat < TEMP_MIN_HABITAT or temp_habitat > TEMP_MAX_HABITAT or
              temp_huerta < TEMP_MIN_HUERTA or temp_huerta > TEMP_MAX_HUERTA):
            return EstadoSistema.ADVERTENCIA
        else:
            return EstadoSistema.NORMAL

    def calcular_potencia_necesaria(self, temp_actual: float, temp_objetivo: float) -> float:
        """Calcula la potencia necesaria para alcanzar la temperatura objetivo"""
        diferencia = temp_objetivo - temp_actual
        potencia = max(0.0, min(100.0, abs(diferencia) * 10.0))
        
        return potencia if diferencia > 0.0 else -potencia

    def ajustar_actuadores(self, zona: ZonaHabitat, potencia: float) -> None:
        """Ajusta los actuadores de una zona con la potencia calculada"""
        actuadores = (self.actuadores_habitat if zona == ZonaHabitat.HABITAT 
                     else self.actuadores_huerta)
        
        actuadores_activos = [a for a in actuadores if a.activo]
        
        if actuadores_activos:
            potencia_por_actuador = abs(potencia) / len(actuadores_activos)
            
            for actuador in actuadores_activos:
                actuador.ajustar_potencia(potencia_por_actuador)

    def protocolo_emergencia(self) -> None:
        """Activa el protocolo de emergencia"""
        print("\n🚨 PROTOCOLO DE EMERGENCIA ACTIVADO")
        print("   1. Activando todos los actuadores disponibles")
        print("   2. Notificando a control de misión")
        print("   3. Preparando sistemas de respaldo")
        
        for actuador in self.actuadores_habitat + self.actuadores_huerta:
            actuador.activar()
            actuador.ajustar_potencia(100.0)

    def ciclo_control(self) -> None:
        """Ejecuta un ciclo completo de control de temperatura"""
        with self._lock:
            print("\n" + "=" * 60)
            print("🔄 Iniciando ciclo de control de temperatura")
            print("=" * 60)

            try:
                # Leer temperaturas
                temp_habitat = self.obtener_temperatura_promedio(ZonaHabitat.HABITAT)
                temp_huerta = self.obtener_temperatura_promedio(ZonaHabitat.HUERTA)

                print("\n📊 LECTURAS ACTUALES:")
                print(f"   Hábitat: {temp_habitat:.2f}°C "
                      f"(Rango: {TEMP_MIN_HABITAT:.1f}°C - {TEMP_MAX_HABITAT:.1f}°C)")
                print(f"   Huerta:  {temp_huerta:.2f}°C "
                      f"(Rango: {TEMP_MIN_HUERTA:.1f}°C - {TEMP_MAX_HUERTA:.1f}°C)")

                # Evaluar estado
                self.estado = self.evaluar_estado(temp_habitat, temp_huerta)
                
                iconos = {
                    EstadoSistema.NORMAL: "✓",
                    EstadoSistema.ADVERTENCIA: "⚠",
                    EstadoSistema.CRITICO: "🔴",
                    EstadoSistema.EMERGENCIA: "🚨"
                }
                
                print(f"\n{iconos[self.estado]} ESTADO DEL SISTEMA: {self.estado.value.upper()}")

                # Calcular y aplicar ajustes
                temp_objetivo_habitat = (TEMP_MIN_HABITAT + TEMP_MAX_HABITAT) / 2.0
                temp_objetivo_huerta = (TEMP_MIN_HUERTA + TEMP_MAX_HUERTA) / 2.0

                potencia_habitat = self.calcular_potencia_necesaria(
                    temp_habitat, temp_objetivo_habitat)
                potencia_huerta = self.calcular_potencia_necesaria(
                    temp_huerta, temp_objetivo_huerta)

                print("\n🎯 AJUSTES CALCULADOS:")
                self.ajustar_actuadores(ZonaHabitat.HABITAT, potencia_habitat)
                self.ajustar_actuadores(ZonaHabitat.HUERTA, potencia_huerta)

                # Protocolo de emergencia
                if self.estado == EstadoSistema.EMERGENCIA:
                    self.protocolo_emergencia()

                print("✓ Ciclo completado exitosamente")

            except Exception as e:
                print(f"✗ Error en ciclo: {e}")
                raise

    def diagnostico_sistema(self) -> None:
        """Muestra un diagnóstico completo del sistema"""
        with self._lock:
            print("\n" + "=" * 60)
            print("🔍 DIAGNÓSTICO DEL SISTEMA")
            print("=" * 60)
            
            print("\n📡 Sensores Hábitat:")
            for sensor in self.sensores_habitat:
                print(f"   {sensor}")
            
            print("\n📡 Sensores Huerta:")
            for sensor in self.sensores_huerta:
                print(f"   {sensor}")
            
            print("\n🔧 Actuadores Hábitat:")
            for actuador in self.actuadores_habitat:
                print(f"   {actuador}")
            
            print("\n🔧 Actuadores Huerta:")
            for actuador in self.actuadores_huerta:
                print(f"   {actuador}")


def main():
    """Función principal del sistema"""
    print("=" * 60)
    print("🚀 SISTEMA DE CONTROL DE TEMPERATURA - HÁBITAT MARCIANO")
    print("=" * 60)
    print("Versión: 1.0.0 - Python")
    print("Estado: Sistema Crítico - Máxima Prioridad")

    try:
        sistema = SistemaControlTemperatura()
        
        # Diagnóstico inicial
        sistema.diagnostico_sistema()
        
        print("\n🔄 Iniciando monitoreo continuo...\n")
        
        # Simulación de 5 ciclos de control
        for ciclo in range(1, 6):
            print("\n" + "=" * 60)
            print(f"CICLO #{ciclo}")
            
            time.sleep(2)
            
            sistema.ciclo_control()
        
        # Diagnóstico final
        sistema.diagnostico_sistema()
        
        print("\n" + "=" * 60)
        print("✓ Sistema de control finalizado")
        print("=" * 60)

    except KeyboardInterrupt:
        print("\n\n⚠ Sistema interrumpido por el usuario")
        sys.exit(0)
    except Exception as e:
        print(f"\n🚨 Error crítico: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

# Made with Bob
